src/main/java/org/cyclonedx/gradle/DependencyGraphTraverser.java

```diff
@@ -0,0 +1,197 @@
+                        final ResolvedComponentResult dependencyComponent =
+                                ((ResolvedDependencyResult) dep).getSelected();
+                        final GraphNode dependencyNode = new GraphNode(dependencyComponent);
+                        dependencyNode.inScopeConfiguration(projectName, configName);
+                        graph.get(graphNode).add(dependencyNode);
+                        queue.add(dependencyNode);
+                    }
+                }
+            }
+        }
+
+        return toSbomComponents(graph);
+    }
+
+    private Map<SbomComponentId, SbomComponent> toSbomComponents(final Map<GraphNode, Set<GraphNode>> graph) {
+        return graph.entrySet().stream()
+                .map(entry -> toSbomComponent(entry.getKey(), entry.getValue()))
+                .collect(Collectors.toMap(v -> v.getId(), v -> v));
+    }
+
+    private SbomComponent toSbomComponent(final GraphNode node, final Set<GraphNode> dependencyNodes) {
+
+        final File artifactFile = getArtifactFile(node);
+        final SbomComponentId id = DependencyUtils.toComponentId(node.getResult(), artifactFile);
+
+        List<License> licenses = null;
+        SbomMetaData metaData = null;
+        if (includeMetaData && node.id instanceof ModuleComponentIdentifier) {
+            final Component component = new Component();
+            extractMetaDataFromArtifactPom(artifactFile, component, node.getResult());
+            licenses = extractMetaDataFromRepository(component, node.getResult());
+            metaData = SbomMetaData.fromComponent(component);
+        }
+
+        return new SbomComponent.Builder()
+                .withId(id)
+                .withDependencyComponents(getSbomDependencies(dependencyNodes))
+                .withInScopeConfigurations(node.getInScopeConfigurations())
+                .withArtifactFile(artifactFile)
+                .withMetaData(metaData)
+                .withLicenses(licenses)
+                .build();
+    }
+
+    private void extractMetaDataFromArtifactPom(
+            final File artifactFile, final Component component, final ResolvedComponentResult result) {
+
+        if (artifactFile == null) {
+            return;
+        }
+
+        final MavenProject mavenProject = mavenHelper.extractPom(artifactFile, result.getModuleVersion());
+        if (mavenProject != null) {
+            mavenHelper.getClosestMetadata(artifactFile, mavenProject, component, result.getModuleVersion());
+        }
+    }
+
+    private List<License> extractMetaDataFromRepository(
+            final Component component, final ResolvedComponentResult result) {
+        final MavenProject mavenProject = mavenLookup.getResolvedMavenProject(result);
+        if (mavenProject != null) {
+            mavenHelper.extractMetadata(mavenProject, component);
+            return mavenProject.getLicenses();
+        }
+
+        return null;
```

Comment:
Please, let's return empty collection instead

Reply:
Yes good point, added that.